src/Microsoft.ML.Data/DataLoadSave/PartitionedFileLoader.cs

```diff
@@ -0,0 +1,737 @@
+
+            [Argument(ArgumentType.Required, HelpText = "Source index of the column.")]
+            public int Source;
+
+            public static Column Parse(string str)
+            {
+                Contracts.AssertNonEmpty(str);
+
+                if (TryParse(str, out Column column))
+                {
+                    return column;
+                }
+
+                return null;
+            }
+
+            public static bool TryParse(string str, out Column column)
+            {
+                column = null;
+
+                if (string.IsNullOrEmpty(str))
+                {
+                    return false;
+                }
+
+                if (!ColumnParsingUtils.TryParse(str, out string name, out string sourceStr, out string kindStr))
+                {
+                    return false;
+                }
+
+                DataKind? kind = null;
+                if (kindStr != null && TypeParsingUtils.TryParseDataKind(kindStr, out DataKind parsedKind, out KeyRange range))
+                {
+                    kind = parsedKind;
+                }
+
+                if (!int.TryParse(sourceStr, out int source))
+                {
+                    return false;
+                }
+
+                column = new Column()
+                {
+                    Name = name,
+                    Source = source,
+                    Type = kind
+                };
+
+                return true;
+            }
+
+            public bool TryUnparse(StringBuilder sb)
+            {
+                Contracts.AssertValue(sb);
+
+                sb.Append($"{Name}");
+
+                if (Type.HasValue)
+                {
+                    sb.Append($":{Type}");
+                }
+
+                sb.Append($":{Source}");
+
+                return true;
+            }
+        }
+
+        private readonly IHost _host;
+        private readonly IMultiStreamSource _files;
+        private readonly Column[] _columns;
+
+        // Number of tailing directories to include.
+        private readonly int _tailingDirCount;
+
+        // An underlying loader used on each individual loader.
+        private readonly SubComponent<IDataLoader, SignatureDataLoader> _subLoader;
+
+        private readonly IPartitionedPathParser _pathParser;
+
+        private const string RegistrationName = LoadName;
+        private const string FilePathSpecName = "FilePathSpec";
+        private const int FilePathColIndex = -1;
+
+        public PartitionedFileLoader(IHostEnvironment env, Arguments args, IMultiStreamSource files)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(RegistrationName);
+            _host.CheckValue(args, nameof(args));
+            _host.CheckValue(files, nameof(files));
+
+            _pathParser = args.PathParserFactory.CreateComponent(_host);
+            _host.CheckValue(_pathParser, nameof(_pathParser), "Factory failed to create a FilePathSpec");
```

Comment:
>CheckValue(_pathParser, nameof(_pathParser) [](start = 18, length = 43)

You cannot use `CheckValue` here (as evidenced by the fact that the parameter name you're passing in is the name of a private field of the class that the user has no idea about). Use `Check(_pathParser != null, "...")` instead with your error message.

Reply:
I'm still seeing `CheckValue` on a private field. Note that `CheckValue` is meant to indicate to the caller of a method that a parameter that shouldn't be null was, as we see since the first parameter is the parameter name. This is not a parameter so it should not be used. Rather, this is a problem with `PathParserFactory`. It should be a `UserArg` style exception, on `nameof(args.PathParserFactory)`.

---
In reply to: [187712097](https://github.com/dotnet/machinelearning/pull/61#discussion_r187712097) [](ancestors = 187712097)

Reply:
Hi @tyclintw do you see this comment?

---
In reply to: [190669939](https://github.com/dotnet/machinelearning/pull/61#discussion_r190669939) [](ancestors = 190669939,187712097)

Reply:
I did not see the comment, thank you.